### PR TITLE
OpenAI Chat & Assistant hook functions

### PR DESCRIPTION
**Pull request by nathadfield**
Adding a set of additional hook functions for the OpenAI provider that deal with interactions between some of the [Chat](https://platform.openai.com/docs/api-reference/chat) and [Assistants](https://platform.openai.com/docs/api-reference/assistants) API endpoints.

### CODE REVIEW

**Comment by nathadfield** [comment]
I'd love a review on this when someone gets an opportunity.  Thanks!

File: airflow/providers/openai/hooks/openai.py [hunk 1]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+        **kwargs: Any,
+    ) -> ChatCompletionMessage:
+        """
+        Create a model response for the given chat conversation and returns the message.
+
+        :param messages: A list of messages comprising the conversation so far
+        :param model: ID of the model to use
+        """
+        response = self.conn.chat.completions.create(model=model, messages=messages, **kwargs)
+        return response.choices[0].message
```
Comment (utkarsharma2) [inline airflow/providers/openai/hooks/openai.py]: Will the response always contain a choices list with one value?

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: It's a good question.  I've not seen any different and this is based upon OpenAI's official [docs](https://platform.openai.com/docs/api-reference/chat).

File: airflow/providers/openai/hooks/openai.py [hunk 2]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+
+    def create_assistant(self, model: str = "gpt-3.5-turbo", **kwargs: Any) -> Assistant:
+        """Create an OpenAI assistant using the given model.
+
+        :param model: The OpenAI model for the assistant to use.
+        """
+        assistant = self.conn.beta.assistants.create(model=model, **kwargs)
```
Comment (utkarsharma2) [inline airflow/providers/openai/hooks/openai.py:118]: Just curious, are we using beta API for OpenAI here?

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py:118]: AFAIK, this is all that's available for the [Assistants API](https://platform.openai.com/docs/api-reference/assistants) currently.

File: airflow/providers/openai/hooks/openai.py (hunk 1 above)
Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: It does actually say that there can be more than one value if the the value of `n` submitted to `chat.completions.create` is more than 1.

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]:
```
n
integer or null

Optional
Defaults to 1
How many chat completion choices to generate for each input message. Note that you will be charged based on the number of generated tokens across all of the choices. Keep n as 1 to minimize costs.
```

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: I could set `n=1` as a default argument which would return `response.choices[0].message` of type `Message` but if this was overwritten it could return `response.choices` which would be `list[Message]` but I'm not sure if that would be a good practice or not.

File: airflow/providers/openai/hooks/openai.py [hunk 3]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+    def get_assistant(self, assistant_id: str) -> Assistant:
+        """
+        Get an OpenAI assistant.
+
+        :param assistant_id: The ID of the assistant to retrieve.
+        """
+        assistant = self.conn.beta.assistants.retrieve(assistant_id=assistant_id)
+        return assistant
+
+    def get_assistants(self, **kwargs) -> list[Assistant]:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def get_assistants(self, **kwargs: Any) -> list[Assistant]:
```

File: airflow/providers/openai/hooks/openai.py [hunk 4]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+        """Get a list of Assistant objects."""
+        assistants = self.conn.beta.assistants.list(**kwargs)
+        return assistants.data
+
+    def get_assistant_with_name(self, assistant_name: str) -> Assistant | None:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def get_assistant_by_name(self, assistant_name: str) -> Assistant | None:
```

Not sure whether it sounds more reasonable 🤔

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: It definitely does!

File: airflow/providers/openai/hooks/openai.py (hunk 1 above)
Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: @Lee-W Do you have any thoughts on this?

Comment (Lee-W) [reply airflow/providers/openai/hooks/openai.py]: I feel like we probably should return `list[Message]`. Will that cause any issue if we do so?

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: I suppose not.  My initial thought was that I just wanted to return the message object but, if there can be more than one, then returning the list and then handling this outside probably seems like the correct thing to do.

**Comment by teowave** [comment]
Excellent and timely addition. I was just struggling to get openai chat completions to work straight from within pythonoperator, only to silently hang. Looking forward to seeing this implemented

File: airflow/providers/openai/hooks/openai.py (hunk 1 above)
Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: Added this change.

**Comment by nathadfield** [comment]
One failed test here but I don't think it's anything to do with my changes.

**Comment by Lee-W** [comment]
> One failed test here but I don't think it's anything to do with my changes.

Just retrigger it. Let's see how it works

**Comment by nathadfield** [comment]
I rebased which ran the checks agan but it's now failing with a different error but still not connected to my changes.  How else can I retrigger the checks?

**Comment by nathadfield** [comment]
@Lee-W I keep rebasing and rerunning but some unrelated tests always seem to fail.  Any suggestions?

**Comment by Lee-W** [comment]
Looks like network issue 🤔 let me try again. if it still fails, I'll take a deeper look

**Comment by nathadfield** [comment]
> Looks like network issue 🤔 let me try again. if it still fails, I'll take a deeper look

You seem to have the magic touch!

**Comment by nathadfield** [comment]
Just looking for someone to give it a 👍 now.

File: airflow/providers/openai/hooks/openai.py [hunk 5]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+        :param assistant_name: The name of the assistant to retrieve
+        """
+        response = self.get_assistants()
+        for assistant in response:
+            if assistant.name == assistant_name:
+                return assistant
+        return None
+
+    def modify_assistant(self, assistant_id: str, **kwargs: Any) -> Assistant:
+        """Modify an existing Assistant object."""
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]: missing `:param assistant_id` in the docstring. let's make it consistent

File: airflow/providers/openai/hooks/openai.py [hunk 6]
```diff
@@ -77,6 +89,165 @@ def get_conn(self) -> OpenAI:
...
+        """
+        response = self.conn.beta.assistants.delete(assistant_id=assistant_id)
+        return response
+
+    def create_thread(self, **kwargs: Any) -> Thread:
+        """Create an OpenAI thread."""
+        thread = self.conn.beta.threads.create(**kwargs)
+        return thread
+
+    def modify_thread(self, thread_id: str, metadata: dict) -> Thread:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]: What is metadata? Should it be typed as `dict[str, Any]`?

File: tests/providers/openai/hooks/test_openai.py [hunk 7]
```diff
@@ -56,6 +69,226 @@ def mock_embeddings_response():
...
+    assert assistant.instructions == ASSISTANT_INSTRUCTIONS
+
+
+def test_get_assistants(mock_openai_hook, mock_assistant_list):
+    mock_openai_hook.conn.beta.assistants.list.return_value = mock_assistant_list
+    assistants = mock_openai_hook.get_assistants()
+    assert isinstance(assistants, list)
+
+
+def test_get_assistant_with_name(mock_openai_hook, mock_assistant_list):
```
Comment (Lee-W) [inline tests/providers/openai/hooks/test_openai.py]:
```suggestion
def test_get_assistant_by_name(mock_openai_hook, mock_assistant_list):
```

**Review by Lee-W — COMMENTED** [review]
Overall, it looks good. Left a few nitpicks. But would like have @utkarsharma2 to take another look to ensure I did not miss anything crucial 🙂

File: airflow/providers/openai/hooks/openai.py (hunk 6 above)
Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: Yes, I think that makes sense.  

The `metadata` is just a set of K:V pairs that can be attached to the object. e.g. 

```
my_updated_thread = client.beta.threads.update(
  "thread_abc123",
  metadata={
    "modified": "true",
    "user": "abc123"
  }
)
```

**Comment by nathadfield** [comment]
@Lee-W No worries.  Thanks for the input.  

@utkarsharma2 It'd be great to get any more thoughts from you.

**Review by Lee-W — APPROVED** [review]

**Comment by nathadfield** [comment]
Is someone able to merge this soon then?

**Comment by Lee-W** [comment]
I'm planning on merging this one later today. Please let me know if anyone want to take a deeper look.

**Review by pankajkoti — APPROVED** [review]
